Add documentation for running a packaged project (#2425)
kedro-org/kedro: docs/source/tutorial/package_a_project.md

The project configuration is provided separately in a `tar.gz` file, also inside the `dist` folder. This compressed version of the config files excludes any files inside the `local` directory.

### Package recipients
### Run a packaged project

Recipients of the `.whl` file need to have Python and `pip` on their machines, but do not need to have Kedro installed.

A recipient can install the project by calling:
To run a packaged project it must first be installed. To install the project, run the following command:

```bash
pip install <path-to-wheel-file>
```

An executable, `spaceflights`, is placed in the `bin` subfolder of the Python install folder, so the project can be run as follows:

```bash

> python -m spaceflights

Reviewer: I am a bit hesitant about removing the mention of `python -m <package>` here.

1. We still have `python -m package` mentioned in a few place
2. This seems to be going to be the way how we deploy kedro, at the moment we have issues with a `click` so it doesn't work well on Databricks, but it's a special case. I personally

Author: Both this way and the way I describe are equivalent. They both call `__main__:main`. In fact, both expose the `click` options for `kedro run`. Try running `python -m <package_name> --help` and checking it for yourself.

I'm not too concerned about Databricks, as this PR also adds clear instructions for running a packaged project on that platform.

That being said, I've added this back in.

> To run your project using its CLI, run `<project_name>`. To see a full list of the options that the project's CLI supports, you can run `<project_name> --help`.

Reviewer 2: Worth adding an example snippet right below?

Author: I thought about this, but it is difficult to justify as the commands are so short. Willing to be overruled but I would keep it as is.

> To run your project using its CLI, run `<project_name>`. To see a full list of the options that the project's CLI supports, you can run `<project_name> --help`.

Reviewer: I am a bit confused about what it means. If we have a project name `spaceflights`, does it mean I should run `spaceflights` in the terminal? Does it actually works?

Author: It works, in `setup.py` we define the default entry point for the script:

```python
entry_point = (
    "{{ cookiecutter.repo_name }} = {{ cookiecutter.python_package }}.__main__:main"
)
```

This entry point allows the user to run the package after installing it by calling the project's name (`cookiecutter.repo_name`).

Author: It might be confusing as I may have the wrong understanding of `package_name` and `project_name`.

Reviewer: Thanks for explaining, I wasn't awared that we register an entrypoint automatically.

Author: I've looked into this more and decided that `<project_name>` is not right here, it would be inconsistent with how we use `<project_name>` elsewhere in the code base (we map it to `{{ cookiecutter.project_name }}`). I have changed this to `<repo_name>` and added a note to the user what is meant by `repo_name`.

I have modified the precedence of both ways of running a packaged project so that your preferred way (`python -m ...` is now first.

The project configuration is provided separately in a `tar.gz` file, also inside the `dist` folder. This compressed version of the config files excludes any files inside the `local` directory.

### Run a packaged project

To run a packaged project it must first be installed. To install the project, run the following command:

```bash
pip install <path-to-wheel-file>
```

Once your project is installed, it  can be run in two ways:

> 1. Using the project's CLI.

Reviewer 3: Numbered bullets imply a sequence, that is, you have to do both. I would instead have an "either X or Y" statement as a sentence rather than a list.

Author: Thanks for explaining this, that makes sense.

> 1. Using the project's CLI.

Author: In my view, there are two things happening:

- The entry point to the script (which points to `package.__main__`) has the name `<package_name>`. It can be used to run the project.
- The entry point uses Kedro's `run` API, which is decorated using `click`. This allows options and arguments passed to the entry point to be consumed, for instance: `my-project --from-nodes A,B`. In my view, this qualifies as a CLI for the packaged project.

I feel that this 'CLI' approach to running a project is also more intuitive to beginner users than using `python`'s `-m` option. What do you think @stichbury?

> I think we should also explain what does python -m does, under the hood it picks up the package.__main__ which user actually has the ability to change before they package it.

I agree with you that this would be generally useful information, but I personally feel that this is too involved for our tutorial. A minority of advanced users may want to modify this behaviour and my view is that they are likely to have sufficient knowledge of Python packaging to achieve this independently. I would avoid overwhelming beginner and intermediate users with information.

2. Interactively using Python code.

> To run the project using its CLI, you can either use the package name (simply `<package_name>` from the command line), or the package name followed by the `run` command (`<package_name> run`). To see a full list of the options that the project's CLI supports, you can run `<package_name> -h`.

Author: You're right that it doesn't expose `run`, it exposes the run API. I made this change in [30462c8](https://github.com/kedro-org/kedro/pull/2425/commits/30462c8f64c193ae89af58117f2b392a86a69008).

This is equivalent to running `kedro run`, and you can provide all the parameters described by `kedro run --help`.

To install the package from a `.whl` file, you need to have Python and `pip` on your machine, but do not need to have Kedro installed.

```{note}
Once the packaged project is installed, you will need to add:

* a `conf` folder
* a `data` folder if the pipeline loads/saves local data

> * a `logs` folder if the project uses logging.
> Alternatively, they can make use of the ``OmegaConfigLoader`` to run the configuration directly from the compressed .tar.gz configuration file by running

Reviewer 2: ```suggestion
* a `logs` folder if the project uses logging.

Alternatively, you can make use of the ``OmegaConfigLoader`` to run the configuration directly from the compressed .tar.gz configuration file by running
```

Fixing whitespace and replacing "they" by "you" (hope I understood this correctly)

Author: Thanks for noticing!

> kedro run --conf-source <path-to-compressed-config>.tar.gz

Reviewer 2: To have code snippets inside fenced blocks, you can do

`````
````{note}
This is a note:

```
$ kedro run
```
````
`````

This is equivalent to running `kedro run`, and you can provide all the parameters described by `kedro run --help`.

> Recipients of the `.egg` and `.whl` files need to have Python and `pip` on their machines, but do not need to have Kedro installed.

Reviewer 3: I think we are removing all mention of `.egg` from the docs, right, @astrojuanlu ?

> Recipients of the `.egg` and `.whl` files need to have Python and `pip` on their machines, but do not need to have Kedro installed.

Reviewer 3: ```suggestion
To install the package from a `.whl` file, you need to have Python and `pip` on your machine, but do not need to have Kedro installed.
```

Reviewer 3: I'd try to remove "recipient" as it's not an intuitive word.

```{note}

> The recipient will need to add a `conf` subfolder. They also need to add `data` and `logs` if the pipeline loads/saves local data or uses logging.

Reviewer 3: ```suggestion
Once the packaged project is installed, you will need to add:

* a `conf` folder
* a `data` folder if the pipeline loads/saves local data 
* a `logs` folder if the project uses logging.
```
